In [1]:
# let's start the process to find the linear regression for multiple variables
# first we'll import the libraries required
import numpy as np

In [5]:
# next step is simply to define the variables that we'll be using
# for multiple regression we'll have multiple features for which we'll be using vectors
x = np.array([[2104, 5, 1, 45], [1416, 3, 2, 40], [852, 2, 1,  35]])
y = np.array([460, 232, 178])

The following arrays/vectors are declared from the table:

| Size (sqft) | Number of Bedrooms  | Number of floors | Age of  Home | Price (1000s dollars)  |   
| ----------------| ------------------- |----------------- |--------------|-------------- |  
| 2104            | 5                   | 1                | 45           | 460           |  
| 1416            | 3                   | 2                | 40           | 232           |  
| 852             | 2                   | 1                | 35           | 178           |  

The equation for a multiple linear regression is given as:
$$ f_{\mathbf{w},b}(\mathbf{x}) =  w_0x_0 + w_1x_1 +... + w_{n-1}x_{n-1} + b $$
$$ where, w_0, w_1, w_2...w_{n-1}$$ and b are the parameters such that w is an n-dimensional vector and b is a scalar.

In [2]:
# for the initial purpose let us randomly define the values of w and b
w = np.array([20, 30, 40, 50])
b = 60

In [8]:
# first we'll be preparing the prediction function that'll be used to predict the price of house
def predict(x, w, b):
  p = np.dot(w, x) + b
  return p

# let us check what the predicted price would be for the first data from the table i.e. [2104, 5, 1, 45]
x_init = x[0,:]
got = predict(x_init,w,b)
print('The value of house according to the selected parameters(w,b) and values(x) is: ', got)

The value of house according to the selected parameters(w,b) and values(x) is:  44580


That's way beyond the actual value. So, we have to have more accurate values of w and b. We'll be getting that from the gradient descent but first let's calculate the value of cost function for all the values given in x.

In [9]:
def cost(x, y, w, b):
  m = x.shape[0]
  cost = 0
  for i in range(m):
    p = np.dot(x[i], w) + b
    cost += p
  cost /= (2*m)
  return cost

In [11]:
# let's find what the cost would be if we use the same old values of parameters w and b
got = cost(x, y, w, b)
print("The total cost is: ", got)

The total cost is:  15680.0


The cost is way too much so hope we'd be able to improve it by finding the most suitable parameter values.
The first step to finding the correct values for w and b is given by the gradient descent for which the formula is given by:
$$\begin{align*} \text{repeat}&\text{ until convergence:} \; \lbrace \newline\;
& w_j = w_j -  \alpha \frac{\partial J(\mathbf{w},b)}{\partial w_j}  \; & \text{for j = 0..n-1}\newline
&b\ \ = b -  \alpha \frac{\partial J(\mathbf{w},b)}{\partial b}  \newline \rbrace
\end{align*}$$


We can see that we have to calculate the value of gradients for this process for which the formulas are:
$$
\begin{align}
\frac{\partial J(\mathbf{w},b)}{\partial w_j}  &= \frac{1}{m} \sum\limits_{i = 0}^{m-1} (f_{\mathbf{w},b}(\mathbf{x}^{(i)}) - y^{(i)})x_{j}^{(i)}  \\
\frac{\partial J(\mathbf{w},b)}{\partial b}  &= \frac{1}{m} \sum\limits_{i = 0}^{m-1} (f_{\mathbf{w},b}(\mathbf{x}^{(i)}) - y^{(i)})
\end{align}
$$

In [19]:
# let's calculate the gradient
def gradient(x, y, w, b):
  m, n = x.shape
  dj_db = 0.
  dj_dw = np.zeros((n,))
  for i in range(m):
    err = (np.dot(w,x[i]) + b) - y[i]
    for j in range(n):
      dj_dw[j] += err*x[i][j]
    dj_db += err
  dj_db /= m
  dj_dw /= m
  return dj_db, dj_dw

# now it's time to calculate the gradient descent function
def gradient_descent(x, y, w, b, gradient, iterations, alpha):
  for i in range(iterations):
    dj_db, dj_dw = gradient(x, y, w, b)
    w -= alpha*dj_dw
    b -= alpha*dj_db
  return b, w

# since the gradient descent function has been completed now let's create some values of w and b and pass them to the gradient descent function
m, n = x.shape
w_init = np.zeros(n,)
b_init = 0.
alpha = 5.0e-7
iterations = 1000
b, w = gradient_descent(x, y, w_init, b_init, gradient, iterations, alpha)
print("The return value of parameters from the function for b and w are: b = ", b, " w = ", w)

The return value of parameters from the function for b and w are: b =  -0.002235407530932535  w =  [ 0.20396569  0.00374919 -0.0112487  -0.0658614 ]


In [13]:
# let's check how accurate these parameters define our examples.
x_init = x[0,:]
got = predict(x_init,w,b)
print('The value of house according to the selected parameters(w,b) and values(x) is: ', got)

The value of house according to the selected parameters(w,b) and values(x) is:  426.18530497189204


In [14]:
# That's pretty close to the actual value i.e. 460. Now let's find the total cost of the model
got = cost(x, y, w, b)
print("The total cost is: ", got)

The total cost is:  147.3034013085119


Since 147.3034 is pretty small in comparison to 15680, we can consider this experiment as a success! And this is thus the procedure to find the linear regression for multiple variables.